### DISCUSSION TITLE
Convergence as a function of resolution

### DISCUSSION THREAD

**Discussion by kalosu** [NanoComp/meep#2330] [Q&A]
Hello again friends from the Meep community,

I am trying to use meep to perform some 2D simulations in cylindrical coordinates for a point like emitter which is located in the middle of a GaAs slab. My end goal is to be able to compute the electric field distributions for its use in another simulation software.

In general I have been able to implement what I have just described but I wanted to perform some sort of convergence study in order to guarantee that my results are not influenced by some sort of numerical artifact (or reflections from the PML layers)

For the source definition, I am using a GaussianSource with a single component along Er, with a central wavelength of 900 nm for which I defined the bandwidth as 1% of the central frequency (1/0.9). Additionally, I have set the cutoff parameter to a value of 10, based on the recommendations given in the tutorial: https://meep.readthedocs.io/en/latest/Python_Tutorials/Local_Density_of_States/#extraction-efficiency-of-a-light-emitting-diode-led

What I have done in order to verify the convergence is to evaluate the cross correlation between the obtained field distributions by using the results from the largest resolution (100 pixels per um) as my reference. You can observe the results from this evaluation in the following graph:

![ex_res_sweep_cutoff_10_PML_2_cross_corr_eval](https://user-images.githubusercontent.com/49510503/205981538-fcf592d9-daa3-4943-8782-8205b3a7cd42.png)

In general, you can observe that the general "trend" is towards the results from the largest resolution (100 pixels per mu) but the curve is not smooth and has different jumps in between which I would not be expecting in the case of something that is converging towards something.

Additionally, I have evaluated the extraction efficiency(again following the tutorial just referred)  from my setup as a function of the grid resolution. You can observe the results from this evaluation in the following graph:

![dip_GaAs_pos_130nm_extract_eff_vs_res](https://user-images.githubusercontent.com/49510503/205982476-bd8db9b7-b71d-4f12-a790-1c92eb1111ce.png)

In this case the jump values are not so large, less than 5% from point to point but still I cannot see any clear "trend" towards a definitely value. 

My main question then becomes: How can I perform a proper convergence study in this case? Is this behavior normal or should I expect some clearer trend towards some definite values? 

I suspect that these changes are due to some internal changes in the time stepping and spectral composition of the pulse as a function of the resolution. Do I need to change some of the additional parameters if I adjust the resolution?? (cutoff, decay threshold, pulse bandwidth)

P.S: I have also performed two more additional tests:
-Changing the thickness of the PML layers
-Changing the extension of my simulation domain (along x, given that I am just simulating a flat "infinitely" extended region) 

For both of these two cases, I did not obtain any difference in my results; which indicates that there is no significant contributions from reflections at the edges of the simulation domain.

**Comment by oskooi** [comment]
> In general, you can observe that the general "trend" is towards the results from the largest resolution (100 pixels per mu) but the curve is not smooth and has different jumps in between which I would not be expecting in the case of something that is converging towards something.

Generally when we perform a convergence study, we repeatedly *double* the key simulation parameters of resolution, run time, and PML thickness. You then plot the relative error in some quantity e.g. extraction efficiency, (with the reference result computed at twice the largest resolution/runtime/PML thickness of your dataset) versus the simulation parameter on a log-log scale. This should hopefully make the trend in the data smoother and its asymptotic limit more noticeable. 

The "noise" that you are seeing in your data may just be due to the relatively small resolution range.
